Use guava cache for instance admin endpoints

jackjlli opened the pull request:
This PR:
* uses guava cache for instance admin endpoints.
* removes cached data accessor, which comes from Helix code.

Current usage of cached based data accessor has the issue described in this link: #4665
And the cached data accessor was to solve the problem of frequent zk access to get the instance admin endpoints when there's segment upload. Plus, we've seen some flaky tests in our build, which are also related to this cached data accessor.
This issue can be solved by just introducing a guava cache to cache the instance admin endpoint. So no need to maintain a cached data accessor any more.

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:
```diff
@@ -117,11 +118,13 @@
 public class PinotHelixResourceManager {
   private static final Logger LOGGER = LoggerFactory.getLogger(PinotHelixResourceManager.class);
   private static final long DEFAULT_EXTERNAL_VIEW_UPDATE_RETRY_INTERVAL_MILLIS = 500L;
+  private static final long CACHE_TIMEOUT_HR = 6L;
```
Actually `CACHE_ENTRY_EXPIRE_TIME_HOURS` is more accurate. We don't want to use refresh as the performance is not that critical, and the entry can be removed.

Jackie-Jiang commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java:
```diff
@@ -173,24 +172,18 @@ private void addAndRemoveNewInstanceConfig(ZkClient zkClient)
-      latestAllInstances = _helixResourceManager.getAllInstances();
-    }
-    Assert.assertTrue(System.currentTimeMillis() < maxTime, "Timeout when waiting for adding instance config");
+    TestUtils.waitForCondition(aVoid -> {
```
We don't need to wait for condition right?

jackjlli replied:
It's possible that direct data accessor still fails to get the instance config because of some other issues like zk connection. Let's still keep the wait for condition here.

Jackie-Jiang commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java:
```diff
@@ -173,24 +172,18 @@ private void addAndRemoveNewInstanceConfig(ZkClient zkClient)
-      latestAllInstances = _helixResourceManager.getAllInstances();
-    }
-    Assert.assertTrue(System.currentTimeMillis() < maxTime, "Timeout when waiting for removing instance config");
+    TestUtils.waitForCondition(aVoid -> {
```
Same here

kishoreg commented:
Can you please give me sometime. Adding caching on top of ZK is not trivial and in the end we will end up implementing Helix caching code in Pinot.

jackjlli commented:
@mcvsubbu The CI build failed even after we removed cached data accessor. Previously the test did have some flakiness issue.
I've refactored the test in the latest push to make the test more robust.

Jackie-Jiang commented:
> Can you please give me sometime. Adding caching on top of ZK is not trivial and in the end we will end up implementing Helix caching code in Pinot.

This PR is reverting the cache introduced in #3633 that caused some problem, so actually we are on the same page

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:
```diff
@@ -304,8 +302,7 @@ private void addInstanceGroupTagIfNeeded() {
   public InstanceConfig getHelixInstanceConfig(@Nonnull String instanceId) {
-    ZNRecord znRecord = _cacheInstanceConfigsDataAccessor.get("/" + instanceId, null, AccessOption.PERSISTENT);
-    return znRecord != null ? new InstanceConfig(znRecord) : null;
+    return _helixAdmin.getInstanceConfig(_helixClusterName, instanceId);
```
This will throw exception when instance does not exist. You might want to use `_helixDataAccessor.getProperty(_keyBuilder.instanceConfig(instanceId))`

jackjlli replied:
Updated. 👍

Jackie-Jiang commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:
```diff
@@ -2223,8 +2220,9 @@ public RebalanceResult rebalanceTable(String tableNameWithType, Configuration re
   public boolean instanceExists(String instanceName) {
-    ZNRecord znRecord = _cacheInstanceConfigsDataAccessor.get("/" + instanceName, null, AccessOption.PERSISTENT);
-    return (znRecord != null);
+    HelixDataAccessor helixDataAccessor = _helixZkManager.getHelixDataAccessor();
```
`return getHelixInstanceConfig(instanceName) != null`?

jackjlli replied:
Done.
